File: C++/Grind75/Week-2/409.longest-palindrome.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=409 lang=cpp
 *
 * [409] Longest Palindrome
 */

// @lc code=start
class Solution
{
public:
    int longestPalindrome(string s)
    {
        unordered_map<char, int> count;
        int sum = 0;

        for (char x : s)
        {
            count[x]++;
        }

        for (auto it = count.begin(); it != count.end(); it++)
        {
            sum += (it->second / 2) * 2;

            if (it->second % 2 != 0 && sum % 2 == 0)
            {
                sum++;
            }
        }
        return sum;
    }
};
// @lc code=end

File: Java/Grind75/Week2/409.longest-palindrome.java
package Grind75.Week2;
import java.util.*;
/*
 * @lc app=leetcode id=409 lang=java
 *
 * [409] Longest Palindrome
 */

// @lc code=start
class Solution {
    public int longestPalindrome(String s) {
        HashMap<Character, Integer> freq = new HashMap<>();
        for (Character ch : s.toCharArray()) {
            freq.put(ch, freq.getOrDefault(ch, 0) + 1);
        }
        int length = 0;
        for (int count : freq.values()) {
            length += (count / 2) * 2;
            if (length % 2 == 0 && count % 2 == 1)
                length++;
        }
        return length;
    }
}
// @lc code=end